File: 05_swap_risk_sensitivity/python/risk.py
import copy

from curve import YieldCurve


def parallel_shift_curve(
    curve,
    shift_bp
):
    """
    Parallel shift the zero curve.

    Parameters
    ----------
    shift_bp : float
        Shift in basis points.
    """

    shifted_rates = (
        curve.zero_rates
        + shift_bp / 10000
    )

    return YieldCurve(
        curve.maturities,
        shifted_rates
    )


def dv01(
    instrument,
    curve
):
    """
    Compute DV01 using bump-and-reprice.

    DV01:
    ------
    Dollar value of 1bp move.
    """

    base_npv = instrument.npv(curve)

    bumped_curve = parallel_shift_curve(
        curve,
        1
    )

    bumped_npv = instrument.npv(
        bumped_curve
    )

    return bumped_npv - base_npv


def pv01(
    instrument,
    curve
):
    """
    PV01 magnitude.
    """

    return abs(
        dv01(
            instrument,
            curve
        )
    )

File: 05_swap_risk_sensitivity/cpp/risk.cpp
#include "risk.hpp"

double dv01(
    InterestRateSwap& swap,
    YieldCurve& curve
){

    double base_npv =
        swap.npv(curve);

    std::vector<double> bumped_rates =
        curve.zero_rates;

    for(double& r : bumped_rates){

        r += 0.0001;
    }

    YieldCurve bumped_curve(
        curve.maturities,
        bumped_rates
    );

    double bumped_npv =
        swap.npv(bumped_curve);

    return bumped_npv
        - base_npv;
}
